[HUDI-6806] Support Spark 3.5.0

CTTY opened the pull request:
### Change Logs

Support Spark 3.5.0

### Impact

No public API changes

### Risk level (write none, low medium or high below)

Medium

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

vinothchandar commented:
awesome ! @CTTY .

@yihua can we figure out how we can integrate with the native spark reader.

yihua commented:
> awesome ! @CTTY .
> 
> @yihua can we figure out how we can integrate with the native spark reader on top.

Yes.  We should be able to use the native Spark parquet reader from the file format.

CTTY commented:
@hudi-bot run azure

CTTY commented on .github/workflows/bot.yml:
```diff
@@ -284,29 +294,33 @@ jobs:
+#          - flinkProfile: 'flink1.13'
+#            sparkProfile: 'spark2.4'
+#            sparkRuntime: 'spark2.4.8'
+
```
Using my personal docker image to test Spark 3.5 specifically, will revert

CTTY commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/DataFrameUtil.scala:
```diff
@@ -31,7 +33,7 @@ object DataFrameUtil {
   def createFromInternalRows(sparkSession: SparkSession, schema:
   StructType, rdd: RDD[InternalRow]): DataFrame = {
-    val logicalPlan = LogicalRDD(schema.toAttributes, rdd)(sparkSession)
+    val logicalPlan = LogicalRDD(SparkAdapterSupport.sparkAdapter.toAttributes(schema), rdd)(sparkSession)
```
StructType.toAttributes was removed in Spark 3.5.0 by [SPARK-44353](https://github.com/apache/spark/pull/41925)

Solution is to switch to use DataTypeUtils.toAttributes

CTTY commented on hudi-common/src/test/java/org/apache/hudi/common/util/TestClusteringUtils.java:
```diff
@@ -107,6 +108,7 @@ public void testClusteringPlanMultipleInstants() throws Exception {
 
   // replacecommit.inflight doesn't have clustering plan.
   // Verify that getClusteringPlan fetches content from corresponding requested file.
+  @Disabled("Will fail due to avro issue AVRO-3789. This is fixed in avro 1.11.3")
```
avro 1.11.2 can't compare empty map types

CTTY replied:
This fails only for avro1.11.2/Spark 3.5, we can add a if condition here

CTTY commented on hudi-integ-test/src/main/java/org/apache/hudi/integ/testsuite/dag/nodes/BaseValidateDatasetNode.java:
```diff
@@ -244,10 +239,6 @@ private Dataset<Row> getInputDf(ExecutionContext context, SparkSession session,
-    return RowEncoder.apply(schema)
-        .resolveAndBind(JavaConverters.asScalaBufferConverter(attributes).asScala().toSeq(),
-            SimpleAnalyzer$.MODULE$);
+    return SparkAdapterSupport$.MODULE$.sparkAdapter().getEncoder(schema);
```
[SPARK-44531](https://github.com/apache/spark/pull/42134) Encoder inference moved elsewhere in Spark 3.5.0

CTTY commented on hudi-spark-datasource/hudi-spark/pom.xml:
```diff
@@ -245,6 +245,12 @@
+      <artifactId>parquet-hadoop-bundle</artifactId>
+      <version>${parquet.version}</version>
+      <scope>provided</scope>
+    </dependency>
```
Added parquet-hadoop-bundle to fix classpath issues
```
java.lang.NoClassDefFoundError: Could not initialize class org.apache.spark.sql.execution.datasources.parquet.ParquetOptions$

    at org.apache.spark.sql.execution.datasources.parquet.ParquetOptions.<init>(ParquetOptions.scala:50)
    at org.apache.spark.sql.execution.datasources.parquet.ParquetOptions.<init>(ParquetOptions.scala:40)
    at org.apache.spark.sql.execution.datasources.parquet.Spark34LegacyHoodieParquetFileFormat.buildReaderWithPartitionValues(Spark34LegacyHoodieParquetFileFormat.scala:150)

CTTY commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/InsertIntoHoodieTableCommand.scala:
```diff
@@ -180,11 +180,13 @@ object InsertIntoHoodieTableCommand extends Logging with ProvidesHoodieConfig wi
+      case ae: AnalysisException if (ae.getMessage().startsWith("[INCOMPATIBLE_DATA_FOR_TABLE.CANNOT_FIND_DATA] Cannot write incompatible data for the table")
+        || ae.getMessage().startsWith("Cannot write incompatible data to table")) =>
+        planUtils.resolveOutputColumns(catalogTable.catalogTableName, sparkAdapter.toAttributes(expectedSchema), query, byName = false, conf)
     }
```
SPARK-42309 Error message changed in Spark 3.5.0

CTTY commented on hudi-spark-datasource/hudi-spark3.5.x/src/main/scala/org/apache/spark/sql/HoodieSpark35CatalystExpressionUtils.scala:
```diff
@@ -0,0 +1,112 @@
+        case DateDiff(_, OrderPreservingTransformation(attrRef)) => Some(attrRef)
+        case FromUnixTime(OrderPreservingTransformation(attrRef), _, _) => Some(attrRef)
+        case FromUTCTimestamp(OrderPreservingTransformation(attrRef), _) => Some(attrRef)
+        case ParseToDate(OrderPreservingTransformation(attrRef), _, _, _) => Some(attrRef)
```
Added a new empty argument due to [SPARK-43779](https://github.com/apache/spark/pull/41298/files), ParseToDate API change

CTTY commented on hudi-spark-datasource/hudi-spark3.5.x/src/main/scala/org/apache/spark/sql/parser/HoodieSpark3_5ExtendedSqlAstBuilder.scala:
```diff
@@ -0,0 +1,3496 @@
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.{First, Last}
+import org.apache.spark.sql.catalyst.parser.ParserUtils.{checkDuplicateClauses, checkDuplicateKeys, entry, escapedIdentifier, operationNotAllowed, source, string, stringWithoutUnescape, validate, withOrigin}
+import org.apache.spark.sql.catalyst.parser.{EnhancedLogicalPlan, ParseException, ParserInterface}
```
[SPARK-44333](https://github.com/apache/spark/pull/41890), EnhancedLogicalPlan moved to a different package

CTTY commented on packaging/hudi-utilities-bundle/pom.xml:
```diff
@@ -122,6 +122,8 @@
                   <include>com.github.ben-manes.caffeine:caffeine</include>
+                  <!-- SPARK-43489 Spark 3.5+ has marked protobuf as provided -->
+                  <include>com.google.protobuf:protobuf-java</include>
                   <include>com.twitter:bijection-avro_${scala.binary.version}</include>
```
This is needed otherwise deltastreamer would fail due to
```
Exception in thread "main" java.lang.NoClassDefFoundError: com/google/protobuf/Message
    at java.lang.Class.getDeclaredMethods0(Native Method)
    at java.lang.Class.privateGetDeclaredMethods(Class.java:2729)

CTTY commented on packaging/hudi-utilities-slim-bundle/pom.xml:
```diff
@@ -109,6 +109,8 @@
                   <include>com.github.davidmoten:guava-mini</include>
                   <include>com.github.davidmoten:hilbert-curve</include>
+                  <!-- SPARK-43489 Spark 3.5+ has marked protobuf as provided -->
+                  <include>com.google.protobuf:protobuf-java</include>
```
This is needed otherwise deltastreamer would fail due to
```
Exception in thread "main" java.lang.NoClassDefFoundError: com/google/protobuf/Message
    at java.lang.Class.getDeclaredMethods0(Native Method)
    at java.lang.Class.privateGetDeclaredMethods(Class.java:2729)

yihua replied:
This should not be included, as the `hudi-utilities-slim-bundle` should be used with `hudi-spark-bundle`, which contains `com.google.protobuf:protobuf-java`?

CTTY replied:
I don't see `hudi-spark-bundle` including `com.google.protobuf:protobuf-java` now. But yeah I agree if protobuf is in spark bundle then we shouldn't include it in the slim bundle

CTTY commented on pom.xml:
```diff
@@ -2507,6 +2508,66 @@
+        </property>
+      </activation>
+    </profile>
+
```
I haven't changed the default Spark3 profile to Spark 3.5

yihua commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala:
```diff
@@ -77,7 +77,16 @@ object HoodieAnalysis extends SparkAdapterSupport {
-      val dataSourceV2ToV1FallbackClass = "org.apache.spark.sql.hudi.analysis.HoodieDataSourceV2ToV1Fallback"
+      val dataSourceV2ToV1FallbackClass = if (HoodieSparkUtils.isSpark3_5)
+        "org.apache.spark.sql.hudi.analysis.HoodieSpark35DataSourceV2ToV1Fallback"
+      else if (HoodieSparkUtils.isSpark3_4)
+        "org.apache.spark.sql.hudi.analysis.HoodieSpark34DataSourceV2ToV1Fallback"
+      else if (HoodieSparkUtils.isSpark3_3)
+        "org.apache.spark.sql.hudi.analysis.HoodieSpark33DataSourceV2ToV1Fallback"
+      else {
+        // Spark 3.2.x
+        "org.apache.spark.sql.hudi.analysis.HoodieSpark32DataSourceV2ToV1Fallback"
+      }
```
Could Spark 3.4 and below share the same fallback class?

yihua replied:
Looks like we have to follow this way.

yihua commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala:
```diff
@@ -108,6 +119,7 @@ object HoodieAnalysis extends SparkAdapterSupport {
         } else {
           throw new IllegalStateException("Unsupported Spark version")
         }
+      }
```
nit: no need of outer brackets.

yihua commented on hudi-spark-datasource/hudi-spark2/pom.xml:
```diff
@@ -197,6 +197,14 @@
       <optional>true</optional>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-core_${scala.binary.version}</artifactId>
+      <version>${spark2.version}</version>
+      <scope>provided</scope>
+      <optional>true</optional>
+    </dependency>
```
Any reason of adding this?

CTTY replied:
I remember seeing some classpath issues but somehow can't find the exact error message. We can try reverting this change

yihua commented on hudi-spark-datasource/hudi-spark3.0.x/pom.xml:
```diff
@@ -157,6 +157,14 @@
       <optional>true</optional>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-core_${scala.binary.version}</artifactId>
+      <version>${spark30.version}</version>
+      <scope>provided</scope>
+      <optional>true</optional>
+    </dependency>
```
similar here and other poms.

yihua commented on hudi-spark-datasource/hudi-spark3.0.x/src/test/java/org/apache/hudi/internal/HoodieBulkInsertInternalWriterTestBase.java:
```diff
@@ -0,0 +1,174 @@
+/**
+ * Base class for TestHoodieBulkInsertDataInternalWriter.
+ */
+public class HoodieBulkInsertInternalWriterTestBase extends HoodieSparkClientTestHarness {
```
Is this a new class you added?

yihua commented on hudi-spark-datasource/hudi-spark3.4.x/src/test/java/org/apache/hudi/internal/HoodieBulkInsertInternalWriterTestBase.java:
```diff
@@ -0,0 +1,174 @@
+/**
+ * Base class for TestHoodieBulkInsertDataInternalWriter.
+ */
+public class HoodieBulkInsertInternalWriterTestBase extends HoodieSparkClientTestHarness {
```
Looks like this class is added for every Spark versions.  Is this due to rebase misses?

CTTY replied:
This is because this class depends on `SparkDatasetTestUtils` and `SparkDatasetTestUtils` can't be put in `hudi-spark-common` anymore. More details can be found in comment of `SparkDatasetTestUtils`:
```
 * Note: This util class can be only used within `hudi-spark<spark_version>` modules because it
 * relies on SparkAdapterSupport to get encoder for different versions of Spark. If used elsewhere this
 * class won't be initialized properly amd could cause ClassNotFoundException or NoClassDefFoundError
 ```

yihua replied:
Got it.  We can still move this test to `hudi-spark-datasource/hudi-spark` module to avoid duplication, since `hudi-spark-datasource/hudi-spark` module relies on a particular `hudi-spark<spark_version>` when building.

yihua replied:
So this test is only for Spark 3 and we cannot move the relevant test classes to `hudi-spark-datasource/hudi-spark`, otherwise Spark 2 may not be built successfully.

CTTY commented on hudi-common/src/test/java/org/apache/hudi/avro/TestHoodieAvroUtils.java:
```diff
@@ -318,7 +318,7 @@ public void testRemoveFields() {
         + "{\"name\": \"timestamp\",\"type\": \"double\"},{\"name\": \"_row_key\", \"type\": \"string\"},"
         + "{\"name\": \"non_pii_col\", \"type\": \"string\"},"
-        + "{\"name\": \"pii_col\", \"type\": \"string\"}]},";
+        + "{\"name\": \"pii_col\", \"type\": \"string\"}]}";
```
No, it didn't fail before. In Avro 1.11.2 they enforce a stricter schema format

yihua commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/hudi/SparkAdapter.scala:
```diff
@@ -214,4 +215,26 @@ trait SparkAdapter extends Serializable {
    */
   def translateFilter(predicate: Expression, supportNestedPredicatePushdown: Boolean = false): Option[Filter]
+
+  /**
+   * SPARK-44353 StructType#toAttributes was removed in Spark 3.5.0
+   * Use DataTypeUtils#toAttributes for Spark 3.5+
+   */
+  def toAttributes(struct: StructType): Seq[Attribute]
+
+  /**
+   * SPARK-43039 FileIndex#PartitionDirectory refactored in Spark 3.5.0
+   */
+  def toFileStatuses(partitionDirs: Seq[PartitionDirectory]): Seq[FileStatus]
+
+  /**
+   * SPARK-43039 FileIndex#PartitionDirectory refactored in Spark 3.5.0
+   */
+  def newPartitionDirectory(internalRow: InternalRow, statuses: Seq[FileStatus]): PartitionDirectory
+
+  /**
+   * SPARK-44531 Encoder inference moved elsewhere in Spark 3.5.0
+   * Mainly used for unit tests
+   */
+  def getEncoder(schema: StructType): ExpressionEncoder[Row]
```
Fixed now.

yihua commented:
Azure CI on master also fails on the fourth task.  Merging this PR.
<img width="1656" alt="Screenshot 2023-11-15 at 21 56 14" src="https://github.com/apache/hudi/assets/2497195/fdcd6011-d90a-4861-a9a4-c21ed62414ce">

yihua commented:
> I didn't find the bundle jar for Spark 3.5 on Maven Central, am I missing something?

Spark 3.5 bundle jar will be added in Hudi 0.15.0 release.

ranwani commented:
@yihua : We need to use Hudi with Spark 3.5. Can you let me know when is Hudi 0.15.0 release planned?

yihua commented:
> @yihua : We need to use Hudi with Spark 3.5. Can you let me know when is Hudi 0.15.0 release planned?

The 0.15.0 release branch is planned to be cut this month once we verify engine integrations.

ranwani commented:
@yihua Any estimated date for the release?

Gatsby-Lee commented:
> none

what is the release date for Hudi 0.15.0?

CTTY commented:
Hi @Gatsby-Lee , it should be released last week: https://github.com/apache/hudi/tree/release-0.15.0

Gatsby-Lee commented:
> Hi @Gatsby-Lee , it was released last week: https://github.com/apache/hudi/tree/release-0.15.0
Oh..
so, the doc has not been updated yet.

CTTY commented:
We upgraded `hive-storage-api` to `2.8.1` in this PR and recently we found that this may cause issues for HoodieStreamer + ORC source, please see: https://issues.apache.org/jira/browse/HUDI-8081

About why we added `hive-storage-api` to `hudi-common` and upgraded `hive-storage-api`:
>* Bunch of Hive dependencies issue
    * [ERROR] /Users/yxchang/code/Aws157Hudi/hudi-common/src/main/java/org/apache/hudi/common/util/OrcUtils.java:[36,45] package org.apache.hadoop.hive.ql.exec.vector does not exist
        [ERROR] /Users/yxchang/code/Aws157Hudi/hudi-common/src/main/java/org/apache/hudi/common/util/OrcUtils.java:[37,45] package org.apache.hadoop.hive.ql.exec.vector does not exist
        [ERROR] /Users/yxchang/code/Aws157Hudi/hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroOrcWriter.java:[37,45] package org.apache.hadoop.hive.ql.exec.vector does not exist
        [ERROR] /Users/yxchang/code/Aws157Hudi/hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroOrcWriter.java:[38,45] package org.apache.hadoop.hive.ql.exec.vector does not exist
        [ERROR] /Users/yxchang/code/Aws157Hudi/hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroOrcWriter.java:[55,17] cannot find symbol
    * Fixed by adding hive-storage-api in hudi-common



> testOrcIteratorReadData
        * [ERROR] testOrcIteratorReadData  Time elapsed: 2.171 s  <<< ERROR!
            java.lang.NoClassDefFoundError: org/apache/hadoop/hive/ql/exec/vector/DateColumnVector
                    at org.apache.orc.TypeDescription.createRowBatch(TypeDescription.java:491)
                    at org.apache.orc.TypeDescription.createRowBatch(TypeDescription.java:525)
                    at 
        * It's because orc 1.9.1 depends on hive-storage-api 2.8.1
        * Upgrading hive-storage-api in hudi-common to 2.8.1 can solve the problem
